### PR TITLE
Test execution cost of ExecutionManager ovmCALL

### PR DESCRIPTION
**Pull request by elenadimitrova**
**Description**
Here we add a set of unit tests measuring the gas consumption of `ExecutionManager.ovmCALL` function. For the purposes of getting the costs for that, relative to the size of the input and return data arguments only (excluding the inner `.call` logic cost) we measure with a simple pure function with no logic that just returns the bytes data it was given, that is the `Helper_ExecutionManager.returnData` function. 

The following are the results from the various input/output byte pairs:

| inputData bytes size | returnData bytes size | ovmCALL gasUsed | ovmDELEGATECALL gasUsed |
|----------------------|-----------------------|---------|---------|
| 32                   | 96                    | 66,238   | 69,026|
| 64                   | 128                   | 67,587   | 70,527 |
| 96                   | 160                   | 68,938   | 72,031 |
| 128                  | 193                   | 70,288   | 73,534 |

In summary for every additional 32 bytes input and 32 bytes of output data `ovmCALL` spends ~1,350 additional gas on top of the base gas (64,888 gas) and `ovmDELEGATECALL` spends ~1,500 additional gas on top of the base gas (67,526 gas).

**Additional context**
We would like to get a better understanding of the gas overhead which the `OVM_ExecutionManager` consumes performing the virtualized ovm OPCODES.  Most importantly, we are interested in the dynamic ones, such as `ovmCALL` and `ovmDELEGATECALL`.
Note that we do memory checks/manipulation on both of these fields within ovmCALL, so it should not just be the raw internal call costs that we end up measuring.

**Metadata**
- Resolves https://linear.app/optimism/issue/OP-565/gas-meter-ovm-execution-in-l2

### CODE REVIEW

File: packages/contracts/test/contracts/OVM/execution/OVM_ExecutionManager.gas-spec.ts [hunk 1]
```diff
@@ -117,5 +148,123 @@ describe('OVM_ExecutionManager gas consumption', () => {
...
+          ]
+        )
+
+        const tx = await Helper_ExecutionManager.ovmCALLHelper(
+          OVM_TX_GAS_LIMIT,
+          Helper_TestRunner.address,
+          encodedStep,
+          OVM_StateManager.address,
+          { gasLimit: RUN_OVM_TEST_GAS }
+        )
```
Comment (maurelian) [inline packages/contracts/test/contracts/OVM/execution/OVM_ExecutionManager.gas-spec.ts:210]: I think it would be very helpful to future collaborators to included a comment explaining these steps does.
Looks to me like: 
1. prepend the selector for `returnData(bytes)` to the inputData
2. encode that as call data to pass to `Helper_TestRunner.runSingleTestStep()`
3. Call the `Helper_ExecutionManager` with `Helper_TestRunner` as target, which call back to the `Helper_ExecutionManager`'s `returnData` function.

**Review by maurelian — CHANGES_REQUESTED** [review]
Looking good, though I think more comments in the test and new helper contract would be helpful. 
I also left a couple other comments inline.

**Review by ben-chain — CHANGES_REQUESTED** [review]
This is a great start!! 😁 I think it needs a few changes--gonna try to sketch out the next steps I see with this one:

Just to confirm, right now, the call path which I understand is occurring is: `EOA => EMHelper.ovmCALLHelper => EMHelper.ovmCALL => Helper_TestRunner.runSingleTestStep => EMHelper.returnData`.

One issue I can see with this current approach is that we are actually passing in a longer total calldata to the helper, including the `returnData` which we want to get back.  In other words, the actual size of the "OVM calldata" (the length of [this argument](https://github.com/ethereum-optimism/optimism/blob/c88c4532a17e9edc2a94e63c00fec470204d35cd/packages/contracts/contracts/optimistic-ethereum/OVM/execution/OVM_ExecutionManager.sol#L589)) is the length of [this variable](https://github.com/ethereum-optimism/optimism/blob/c88c4532a17e9edc2a94e63c00fec470204d35cd/packages/contracts/test/contracts/OVM/execution/OVM_ExecutionManager.gas-spec.ts#L191), which is a large ABI-encoded `TestStep` struct, not the `dataVariant.inputData.length`.

I would recommend making a helper contract with a function like `returnDataOfSize(uint _sizeOfDataToReturn, bytes calldata _extraUnusedDataToPadLength)` which dynamically generates some empty returndata of the length we want, and accepts (but does not use) arbitrary extra data to make the calldata the length we want.  This approach will also allow you to keep the "return" step inside the helper, instead of adding a call back into the `Helper_ExecutionManager`.

I would also recommend using a new contract instead of the `Helper_TestRunner` because there is a lot of bloat in there which pertains only to the `ExecutionManagerTestRunner`. In general that contract was written to check *what data is returned by the EM* (for example, "does ovmSLOAD return the right value?"  this is why we have `testStep.expectedReturnData`). The current contract bloats gas cost; for instance, right now we are running several `keccak256` operations in `Helper_TestRunner`.  This is contributing to the `receipt.gasUsed` even though we are interested only in the cost incurred by the EM itself.  Ideally we would explicitly account for the gas spent by the helper and ignore it, but until then, we should still minimize that amount.

We are also currently not accounting for the intrinsic gas since the gas is retrieved from a transaction receipt.  We could either subtract that out, or alternatively, have the `Helper_ExecutionManager` measure `gasLeft()` before and after and return that value.

To summarize, these proposed modifications would make the call trace look more like: `EOA => Helper_ExecutionManager.measureGasCost(...) => ExecutionManager.ovmCALL => Helper_DataReturner.returnDataOfSize(...)`.

File: packages/contracts/contracts/test-helpers/Helper_TestRunner.sol [hunk 2]
```diff
@@ -17,10 +17,7 @@ contract Helper_TestRunner {
         bool onlyValidateFlag;
     }
 
-    function runSingleTestStep(
-        TestStep memory _step
-    )
-        public
+    function runSingleTestStep(TestStep memory _step) public
```
[the remark this answers is not in the document]
Comment (elenadimitrova) [reply packages/contracts/contracts/test-helpers/Helper_TestRunner.sol:20]: Happy to have that conversation. Will you arrange as part of the linter introduction work? I feel very strongly against the current default and I doubt I'll ever get used to it.

**Comment by elenadimitrova** [comment]
This has been paused due to planned OVM rework.
